prov/efa/src/efa_mr.c

```diff
@@ -547,6 +550,139 @@ static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr
+ * @return	0 on success.
+ * 		negative libfabric error code on failure.
+ */
+#if HAVE_CUDA
+static
+int efa_mr_update_domain_mr_map(struct efa_mr *efa_mr, struct fi_mr_attr *mr_attr)
+{
+	struct fid_mr *existing_mr_fid;
+	struct efa_mr *existing_mr;
+	CUmemorytype cuda_memory_type;
+	int err;
+
+	mr_attr->requested_key = efa_mr->mr_fid.key;
+	ofi_genlock_lock(&efa_mr->domain->util_domain.lock);
+	err = ofi_mr_map_insert(&efa_mr->domain->util_domain.mr_map, mr_attr,
+				&efa_mr->mr_fid.key, &efa_mr->mr_fid);
+	ofi_genlock_unlock(&efa_mr->domain->util_domain.lock);
+
+	/* FI_ENOKEY can happen legitimately when the key was assigned a CUDA memory region, and the memory region has been freed
+	 * after the registration.
+	 *
+	 * This is because of the the combination of following 3 things:
+	 *  1) cuda memory uses memory cache.
+	 *  2) cuda memory cache's monitor is not monitoring call to cudaFree().
+	 *  3) kernel released the device part of memory registration when cudaFree() is called on a registered cuda memory.
+	 *
+	 * Therefore, when application call cudaFree() on a registered CUDA memory region, the device released the device
+	 * part of the memory region, make the key available for reuse. However, libfabric is not aware of the incident,
+	 * and kept the user space memory registration in its MR cache and MR map.
+	 *
+	 * The long term solution to handle this issue is to use a CUDA memory minitor that monitors call to cudaFree()
+	 * (cuda memhook monitor).
+	 * Before that is implemented, we deregister the staled user space memory registration and retry the insert.
+	 *
+	 * TODO: remove this branch once CUDA memhook monitor is implemented and used by efa.
+	 */
+	if (err == -FI_ENOKEY) {
+		existing_mr_fid = ofi_mr_map_get(&efa_mr->domain->util_domain.mr_map, efa_mr->mr_fid.key);
+		assert(existing_mr_fid);
+		existing_mr = container_of(existing_mr_fid, struct efa_mr, mr_fid);
+
+		if (existing_mr->peer.iface != FI_HMEM_CUDA) {
+			EFA_WARN(FI_LOG_DOMAIN, "key %ld already assigned to buffer: %p hmem_iface: %d length: %ld\n",
+				 existing_mr->mr_fid.key,
+				 existing_mr->ibv_mr->addr,
+				 existing_mr->peer.iface,
+				 existing_mr->ibv_mr->length);
+			return -FI_ENOKEY;
+		}
+
+		assert(existing_mr->peer.iface == FI_HMEM_CUDA);
+		err = ofi_cuPointerGetAttribute(&cuda_memory_type, CU_POINTER_ATTRIBUTE_MEMORY_TYPE,
+						(CUdeviceptr)existing_mr->ibv_mr->addr);
+		if (!err) {
+			EFA_WARN(FI_LOG_DOMAIN, "key %ld already assigned to cuda buffer: %p cuda_memor_type: %d length: %ld\n",
+				 existing_mr->mr_fid.key,
+				 existing_mr->ibv_mr->addr,
+				 cuda_memory_type,
+				 existing_mr->ibv_mr->length);
+			return -FI_ENOKEY;
+		}
+
+		EFA_INFO(FI_LOG_DOMAIN, "key %ld has been assigned to cuda buffer: %p length: %ld, which has since been freed\n",
+			 existing_mr->mr_fid.key,
+			 existing_mr->ibv_mr->addr,
+			 existing_mr->ibv_mr->length);
+
+		assert(efa_mr->domain->cache);
+		pthread_rwlock_rdlock(&mm_list_rwlock);
```

Comment:
I don't think you need these locks...IIRC these locks are to protect the monitors, not the cache

```
/* The read-write lock is an additional lock used to protect the dlist_entry
 * list of ofi_mem_monitor. Due to the necessity of releasing the mm_lock
 * while walking the dlist in ofi_monitor_notify, we need a separate lock to
 * ensure thread safety. This must be a read-write lock because
 * ofi_monitor_notify may be recursive and cannot block multiple walks from
 * occurring at the same time.
 ```
The requirement to calling ofi_mr_cache_notify is:

```
/* Caller must hold ofi_mem_monitor lock as well as unsubscribe from the region */
```

Reply:
but `ofi_mr_cache_notify` simply call `ofi_mr_cache_uncache_entry`, which does not have any lock in it.

Reply:
Yes, when I said you don't need these locks, I meant you don't need the mm_list_rwlock. The requirement to calling ofi_mr_cache_notify is to hold the ofi_mem_monitor_lock, which is mm_lock <-- which you do need.

I'm saying you can remove the rwlock locks.

Reply:
Thank you!
I removed the `rwlock` in latest revision.